viadot/sources/salesforce.py

```diff
@@ -0,0 +1,238 @@
+                that are sent to the Salesforce API for processing. Defaults to 10000.
+            raise_on_error (bool, optional): Whether to raise an exception if a row upsert fails.
+                If False, we only display a warning. Defaults to False.
+        """
+        if df.empty:
+            self.logger.info("No data to upsert.")
+            return
+
+        if external_id_column and external_id_column not in df.columns:
+            raise ValueError(
+                f"Passed DataFrame does not contain column '{external_id_column}'."
+            )
+        records = df.to_dict("records")
+
+        try:
+            response_code = self.salesforce.bulk.__getattr__(table).upsert(
+                data=records,
+                external_id_field=external_id_column,
+                batch_size=batch_size,
+            )
+        except SalesforceMalformedRequest as e:
+            # Bulk insert didn't work at all.
+            raise ValueError(f"Upsert of records failed: {e}") from e
+
+        self.logger.info(f"Successfully upserted bulk records.")
+
+        if any(result.get("success") is not True for result in response_code):
+            # Upsert of some individual records failed.
+            failed_records = [
+                result for result in response_code if result.get("success") is not True
+            ]
+            msg = f"Upsert failed for records {failed_records} with response code {response_code}"
+            if raise_on_error:
+                raise ValueError(msg)
+            else:
+                self.logger.warning(msg)
+
+        self.logger.info(
+            f"Successfully upserted {len(records)} records into table '{table}'."
+        )
+
+    def download(
+        self, query: str = None, table: str = None, columns: List[str] = None
+    ) -> List[OrderedDict]:
+        """
+        Download all data from the indicated table or the result of the specified query.
+
+        Args:
+            query (str, optional): Query for download the specific data. Defaults to None.
+            table (str, optional): Table name. Defaults to None.
+            columns (List[str], optional): List of columns which are needed,
+                requires table argument. Defaults to None.
+
+        Returns:
+            List[OrderedDict]: Selected rows from Salesforce.
+        """
+        if not query:
+            if columns:
+                columns_str = ", ".join(columns)
+            else:
+                columns_str = "FIELDS(STANDARD)"
+            query = f"SELECT {columns_str} FROM {table}"
+        records = self.salesforce.query(query).get("records")
+        # Remove metadata from the data
+        _ = [record.pop("attributes") for record in records]
+        return records
+
+    def to_df(
+        self,
+        query: str = None,
+        table: str = None,
+        columns: List[str] = None,
+    ) -> pd.DataFrame:
+        """
+        Downloads the indicated data and returns the Dataframe.
```

Comment:
Downloads the indicated data returns it as a pandas `DataFrame`

Reply:
[📝 Updated docstrings in Salesforce source](https://github.com/dyvenia/viadot/pull/629/commits/03d07c4da6508d2b436bf0a4bb09e4f7c3a3db54)